### PR TITLE
send additional properties with Marketplace Suggestions tracks events

### PR DESCRIPTION
**Pull request by haszari**
### All Submissions:

* [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:
Adds some additional context info to Marketplace Suggestions tracks events.

### How to test the changes in this Pull Request:
- This PR depends on changes to the suggestion data from WCCOM – this PR is in progress (`#5738`). This PR should be tested with the updated data.
- Checkout this branch and get it set up as the WooCommerce plugin in your development environment. You may need to delete cookies, reset dismissed suggestions, and update the cached suggestion data.
- Log in as a store administrator and view suggestions & interact with suggestions – click the call-to-action, dismiss suggestions.
- Tracks events should have appropriate values for `context target product promoted`.

### CODE REVIEW

File: assets/js/admin/marketplace-suggestions.js [hunk 1]
```diff
@@ -49,20 +49,24 @@
 			}
 
 			window.wcTracks.recordEvent( 'marketplace_suggestion_dismissed', {
-				suggestion_slug: suggestionSlug
+				suggestion_slug: suggestionSlug,
+				context: context,
+				product: product,
```
Comment (timmyc) [inline assets/js/admin/marketplace-suggestions.js]: If for some reason `product`, `promoted`, or `url` are undefined/null do we still want to record the track?

File: assets/js/admin/marketplace-suggestions.js [hunk 2]
```diff
@@ -267,7 +275,7 @@
 
 			// hide promos for things the user already has installed
 			promos = _.filter( promos, function( promo ) {
-				return ! _.contains( marketplace_suggestions.active_plugins, promo['hide-if-active'] );
+				return ! _.contains( marketplace_suggestions.active_plugins, promo['product'] );
```
Comment (timmyc) [inline assets/js/admin/marketplace-suggestions.js]: Is this a different bug fix, not seeing how it is related to the change of the track props.

Comment (timmyc) [reply assets/js/admin/marketplace-suggestions.js]: nm looked at the related json changes and answered the question.

**Review by timmyc — COMMENTED** [review]
Couple of questions, but will be happy to test once the PR this depends on is merged/launched.

File: assets/js/admin/marketplace-suggestions.js (hunk 1 above)
Comment (haszari) [reply assets/js/admin/marketplace-suggestions.js]: Yes we do! These are essentially optional attributes that are supplied from the json end. We want to be robust to them missing for whatever reason.

Even just the fact that the event occurred (e.g. `suggestion_dismissed`) is useful data.

**Comment by haszari** [comment]
> ...happy to test once the PR this depends on is merged/launched.

Thanks @timmyc ! The API changes have been merged and are now launched. Ping me if you hit any issues getting it set up or getting the cached suggestion data updated :)

**Comment by timmyc** [comment]
@haszari testing this out now and seeing the changes coming through, but it appears in all instances ( empty orders, products, and product edit page ) quite a few of the props are undefined:

Note the _undefined_ on product, promoted, and target props here:

`http://pixel.wp.com/t.gif?suggestion_slug=orders-empty-header&context=orders-list-empty-header&product=undefined&promoted=undefined&target=undefined&url=http://one.wordpress.test&products_count=0&orders_gross=0&_en=wcadmin_marketplace_suggestion_displayed&_ui=woo:1xRWxviC+bQx/0mDmyQtxRFb&_ut=anon&_ts=1553538576410&_tz=7&_lg=en-US&_pf=MacIntel&_ht=1440&_wd=2560&_sx=0&_sy=0&_dl=http://one.wordpress.test/wp-admin/edit.php?post_type=shop_order&_dr=http://one.wordpress.test/wp-admin/&_rt=1553538576411&_=_`

Is this expected? Tested just now, so not sure if all required updates have fully be deployed on the JSON side of things.

**Comment by haszari** [comment]
> @haszari testing this out now and seeing the changes coming through, but it appears in all instances ( empty orders, products, and product edit page ) quite a few of the props are `undefined`:

That's not ideal, thanks for picking it up. Now when these are undefined/falsy it will default to empty string for all these new optional props.

> Tested just now, so not sure if all required updates have fully be deployed on the JSON side of things.

Looks like your test site had old data. I've set up a testing site (you should have an email invite to it now) with the latest code.  Here are a few examples from the orders empty-state suggestions:

Screen Shot 2019-03-26 at 10 07 27 AM

Screen Shot 2019-03-26 at 10 07 37 AM

Screen Shot 2019-03-26 at 10 07 53 AM

I'm thinking we should probably check in @psealock and other tracks people to check if we need to prefix our props with `suggestion_` or something. If core tracks is using `product` it might be a little confusing. @psealock can you advise if there are any issues/clashes with the following tracks props for suggestions-related events:

- `context`
- `product`
- `promoted`
- `target`

**Comment by timmyc** [comment]
@haszari 

> Looks like your test site had old data. I've set up a testing site (you should have an email invite to it now) with the latest code. Here are a few examples from the orders empty-state suggestions:

Ah is the data cached locally? Can I clear out a transient or something to just test with my normal site?

**Comment by haszari** [comment]
> Ah is the data cached locally? Can I clear out a transient or something to just test with my normal site?

Yes it's cached in an option, and updated via a `WP_Queue` job.

To delete the cached copy: `wp option delete woocommerce_marketplace_suggestions`.

To trigger an update manually:

- enter `wp shell`
- grab the queue interface `$queue = WC()->queue();`
- add our update job`$queue->add( 'woocommerce_update_marketplace_suggestions' );`
- exit shell
- then trigger a queue update `wp cron event run action_scheduler_run_queue`

It's a bit involved to test manually, but it's lean and automatic in normal use :)

**Comment by psealock** [comment]
> @psealock can you advise if there are any issues/clashes with the following tracks props for suggestions-related events:

Those property names are fine. Since they are properties of that event, they are unique to that event. You could even drop the `suggestion_` from `suggestion_slug` if you want.

**Review by timmyc — APPROVED** [review]
Thanks for the updated cache clearing instructions. Verified that things work well when no suggestions exist, and the new props are being sent as expected.

**Pull request merged**